[Uid] add command to generate UIDs

guillbdx opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | #36102
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!-- required for new features -->

Command to generate UUID and ULID.

Commit:
Command to generate UIDs.

javiereguiluz commented:
@guillbdx thanks for this and all your other contributions related to Uids.

I'd like to ask something: is this really a common need? Do people generate Uids often in the console?

sstok commented:
FYI. Ramsey/uuid has a console library for the UUID component. I runs on Symfony, but I stopped using it because it kept being outdated.

It's mostly usable during development when you need it a value for your tests or fixtures.

MichaelDemeyer commented:
I usually use website to generate uuid when needed. I think it will be slower to use command line 🤔

tarlepp commented:
I have always used that https://www.uuidgenerator.net/ to generate some UUIDs when needed, I don't think that I would never use cli command for that.

stof commented:
@wouterj for hashing password, there is a good reason not to use a third-party site: that would require that you give them the plaintext password.
Another reason is that the command is hashing password with the config used in the project.

For the UUID generation, I'm not sure having a command for it in the core is worth it (there is no configuration needing to match).

nicolas-grekas commented:
https://www.uuidgenerator.net/ doesn't support eg. ULID

derrabus commented on src/Symfony/Bundle/FrameworkBundle/Command/UidGenerateCommand.php:
```diff
@@ -0,0 +1,139 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Command;
```
Can we move this class to the `Uid` component?

derrabus commented on src/Symfony/Bundle/FrameworkBundle/Command/UidGenerateCommand.php:
```diff
@@ -0,0 +1,139 @@
+    private const UUID_6 = '6';
+    private const ULID = 'ulid';
+
+    private static $types = [
```
```suggestion
    private const TYPES = [
```

derrabus commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/console.xml:
```diff
@@ -174,6 +174,10 @@
 
+        <service id="console.command.uid_generate" class="Symfony\Bundle\FrameworkBundle\Command\UidGenerateCommand">
+            <tag name="console.command" command="uid:generate" />
+        </service>
```
I'd only wire this command if the component is available.

guillbdx replied:
Yes, i've updated my code to do so. I've followed the way it's done for other commands: i still set the command in `console.xml`, but i remove the definition in the `FrameworkExtension.php` file if the Uid component is not installed (ie: if `AbstractUid` class doesn't exist).

nicolas-grekas commented:
Oh, another reason to have the command: the ability to represent uids in base58/base32 (aka shorter identifiers)

derrabus commented on src/Symfony/Bundle/FrameworkBundle/Command/UidGenerateCommand.php:
```diff
@@ -0,0 +1,139 @@
+                break;
+        }
+
+        $io->title('Generated UID:');
```
Just a thought: I understand that you want to provide a friendly user interface. However, this command could be consumed by shell scripts if we provided a way to produce a UUID without any additional interaction or output.

wouterj replied:
Maybe we should (if that possible) output everything except from the UID to stderr instead of stdout?

guillbdx replied:
I've updated the code, so it just outputs the result now.

Taluu commented on src/Symfony/Bundle/FrameworkBundle/Command/UidGenerateCommand.php:
```diff
@@ -0,0 +1,139 @@
+
+        $this
+            ->setDefinition([
+                new InputArgument('type', InputArgument::OPTIONAL, 'The type/version of the generated UID.', null),
```
why is this optionnal if you don't allow the null value below ?

guillbdx replied:
Indeed, this part was confusing, and the interactive stuff was a bit needless. I've updated the code: now, the `type` argument is require.

Taluu commented on src/Symfony/Bundle/FrameworkBundle/Command/UidGenerateCommand.php:
```diff
@@ -0,0 +1,139 @@
+Examples:
+
+  <info>php %command.full_name% ulid</info> for generating a ULID.
+  <info>php %command.full_name% 1</info> for generating a UUID in version 1.
```
how about `uuid-1` instead ? `1` isn't self-explanatory

guillbdx replied:
Done.

Taluu reviewed (commented):
I think having an interactive way of covering the `null` defaults isn't such a good idea TBH

Commit:
Remove command if UID component is not installed.

Commit:
Change values from 1, 3, etc to uuid-1, uuid-3, etc.

Commit:
Options for base 32 and 58.

Commit:
Improve help.

Commit:
Tests in base 32 and 58.

Commit:
CS fix

guillbdx commented:
> Oh, another reason to have the command: the ability to represent uids in base58/base32 (aka shorter identifiers) - this needs to be added to the command btw.

@nicolas-grekas ok, i've added base32 and base58 as options.

guillbdx commented on src/Symfony/Bundle/FrameworkBundle/Command/UidGenerateCommand.php:
```diff
@@ -0,0 +1,139 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Command;
```
The Uid component knows nothing about application and commands. Common way to add commands in the Symfony stack is to place them in the FWB: for example the `config:dump` command is in FWB, not in the Config component, `workflow:dump` is in FWB, not in the Workflow component, etc.

nicolas-grekas replied:
I think @derrabus is right, we did this already for a few components, see e.g LintCommand in Yaml

guillbdx replied:
Indeed. I moved it.

Commit:
constant TYPES

guillbdx commented on src/Symfony/Bundle/FrameworkBundle/Command/UidGenerateCommand.php:
```diff
@@ -0,0 +1,139 @@
+    private const UUID_6 = '6';
+    private const ULID = 'ulid';
+
+    private static $types = [
```
Updated, thanks.

guillbdx commented:
Hi, i've updated my PR accordingly to your feedbacks. Main changes are:

- the definition is removed is `Uid` component is not installed
- options `base32` and `base58` to represent the result in base32 or base58.
- use `uuid-1`, `uuid-3`, etc instead of `1`, `3` etc
- removed the interactive stuff and made the `type` argument required.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -345,6 +346,10 @@ public function load(array $configs, ContainerBuilder $container)
         }
 
+        if (!class_exists(AbstractUid::class)) {
+            $container->removeDefinition('console.command.uid_generate');
```
I'd suggest moving this next to line 188 where there are similar checks

guillbdx replied:
Done.

Commit:
Move removing of definition.

Commit:
Move command to Uid Component.

Commit:
Move UidGenerateCommandTest to Uid Component.

Commit:
Uid Component Changelog.

ro0NL commented on src/Symfony/Component/Uid/Command/UidGenerateCommand.php:
```diff
@@ -0,0 +1,125 @@
+                $uid = new Ulid();
+                break;
+            default:
+                throw new InvalidArgumentException('Invalaible UID type. Available values are '.implode(', ', self::TYPES).'.');
```
typo `Invalaible`

guillbdx replied:
Fixed.

ro0NL commented on src/Symfony/Component/Uid/Command/UidGenerateCommand.php:
```diff
@@ -0,0 +1,125 @@
+                break;
+            default:
+                throw new InvalidArgumentException('Invalaible UID type. Available values are '.implode(', ', self::TYPES).'.');
+                break;
```
not really needed

guillbdx replied:
Yes. Removed.

ro0NL commented on src/Symfony/Component/Uid/Command/UidGenerateCommand.php:
```diff
@@ -0,0 +1,125 @@
+        $name = $input->getArgument('name');
+
+        if (\in_array($type, [self::UUID_3, self::UUID_5], true) && !Uuid::isValid($namespace)) {
+            throw new InvalidArgumentException('You must specify a valid namespace as a second argument.');
```
worth generating/outputting  one? perhaps toggle name/namespace args

guillbdx replied:
Do you mean that, in case the user hasn't provided a namespace, we should generate/output a random one?

ro0NL commented on src/Symfony/Component/Uid/Command/UidGenerateCommand.php:
```diff
@@ -0,0 +1,125 @@
+            $uid = $uid->toBase32();
+        }
+
+        if ($input->getOption('base58')) {
```
`elseif`?

guillbdx replied:
Yes, done.

derrabus commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -187,6 +188,9 @@ public function load(array $configs, ContainerBuilder $container)
             if (!class_exists(BaseYamlLintCommand::class)) {
                 $container->removeDefinition('console.command.yaml_lint');
             }
+            if (!class_exists(AbstractUid::class)) {
```
```suggestion
            if (!class_exists(UidGenerateCommand::class)) {
```

💡  Now that the command has moved, we can check for the actual command class instead.

guillbdx replied:
True, fixed.

derrabus commented on src/Symfony/Component/Uid/Command/UidGenerateCommand.php:
```diff
@@ -0,0 +1,125 @@
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \LogicException
```
This exception is not thrown anymore, is it?

guillbdx replied:
Yes. Removed ;)

Commit:
Minor fixes.

Commit:
CS fix

derrabus commented:
To get the tests green, use PHPUnit's `TestCase` class instead of the one from FrameworkBundle and add `symfony/console` to the `require-dev` section of the Uid component.

fabpot commented:
I'd like to add one more reason to NOT have this command: cluttering the CLI output. We've work hard in the past to get rid of commands (by way of removing optional deps) so that the "default" list of commands is at short as possible. I'm leaning towards a :-1: for adding this one.

nicolas-grekas commented:
Let's close then, time we tell if we really miss this.
